android/app/src/main/java/com/coboltforge/dontmind/multivnc/PointerInputHandler.java

```diff
@@ -432,66 +433,60 @@ private boolean spenActionConvert(MotionEvent e) {
-                } else if ((e.getButtonState() & MotionEvent.BUTTON_SECONDARY) != 0) {
-                    button = true;
-                    secondary = true;
-                    action = MotionEvent.ACTION_DOWN;
-                }
-                if (e.getAction() == MotionEvent.ACTION_MOVE) {
-                    action = MotionEvent.ACTION_MOVE;
-                }
-            } else if ((e.getActionMasked() == MotionEvent.ACTION_HOVER_MOVE) ||
-                    (e.getActionMasked() == MotionEvent.ACTION_UP)){
-                action = MotionEvent.ACTION_UP;
-                button = false;
-                secondary = false;
-            }
+        vncCanvasActivity.vncCanvas.changeTouchCoordinatesToFullFrame(e);
+        int x = (int) e.getX();
+        int y = (int) e.getY();
 
-            if (action != MotionEvent.ACTION_MASK) {
-                e.setAction(action);
-                if (!button) {
-                    vncCanvasActivity.vncCanvas.processPointerEvent(e, false);
-                }
-                else {
-                    vncCanvasActivity.vncCanvas.processPointerEvent(e, true, secondary);
-                }
-                vncCanvasActivity.vncCanvas.panToMouse();
-            }
+        switch (e.getAction()) {
+            case MotionEvent.ACTION_HOVER_MOVE:
+                vncCanvasActivity.vncCanvas.processMouseEvent(0, false, x, y);
+                break;
 
-            if(e.getAction() == MotionEvent.ACTION_SCROLL) {
+            case MotionEvent.ACTION_SCROLL:
                 if (e.getAxisValue(MotionEvent.AXIS_VSCROLL) < 0.0f) {
-                    if (Utils.DEBUG()) Log.d(TAG, "Input: scroll down");
-                    vncCanvasActivity.vncCanvas.vncConn.sendPointerEvent(vncCanvasActivity.vncCanvas.mouseX, vncCanvasActivity.vncCanvas.mouseY, e.getMetaState(), VNCConn.MOUSE_BUTTON_SCROLL_DOWN);
+                    vncCanvasActivity.vncCanvas.processMouseEvent(VNCConn.MOUSE_BUTTON_SCROLL_DOWN, true, x, y);
+                    vncCanvasActivity.vncCanvas.processMouseEvent(VNCConn.MOUSE_BUTTON_SCROLL_DOWN, false, x, y);
+                } else {
+                    vncCanvasActivity.vncCanvas.processMouseEvent(VNCConn.MOUSE_BUTTON_SCROLL_UP, true, x, y);
+                    vncCanvasActivity.vncCanvas.processMouseEvent(VNCConn.MOUSE_BUTTON_SCROLL_UP, false, x, y);
                 }
-                else {
-                    if (Utils.DEBUG()) Log.d(TAG, "Input: scroll up");
-                    vncCanvasActivity.vncCanvas.vncConn.sendPointerEvent(vncCanvasActivity.vncCanvas.mouseX, vncCanvasActivity.vncCanvas.mouseY, e.getMetaState(), VNCConn.MOUSE_BUTTON_SCROLL_UP);
-                }
-            }
+                break;
+        }
 
-        if(Utils.DEBUG())
-            Log.d(TAG, "Input: generic motion: x:" + e.getX() + " y:" + e.getY() + " action:" + e.getAction() + " button:" + button + " secondary:" + secondary);
+        if (Build.VERSION.SDK_INT >= 23) {
```

Comment:
just curious: what happens with SDK_INT < 23 with this code?

Reply:
For < 23, Android only sends `ACTION_UP` & `ACTION_DOWN` events mouse button clicks. These are handled as before in `onTouchEvent()`.